src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,293 @@
+                        {
+                            ExpressionSyntax GetLeftmostCondition(ExpressionSyntax syntaxNode)
+                            {
+                                syntaxNode = syntaxNode.WalkDownParentheses();
+                                while (syntaxNode.IsKind(SyntaxKind.LogicalAndExpression))
+                                {
+                                    syntaxNode = ((BinaryExpressionSyntax)syntaxNode).Left.WalkDownParentheses();
+                                }
+
+                                return syntaxNode;
+                            }
+
+                            var node = (BinaryExpressionSyntax)operand;
+                            var leftmost = GetLeftmostCondition(node);
+                            switch (leftmost.Kind())
+                            {
+                                case SyntaxKind.IsPatternExpression:
+                                case SyntaxKind.IsExpression:
+                                case SyntaxKind.EqualsExpression:
+                                    break;
+
+                                default:
+                                    return null;
+                            }
+
+                            var p = GetPattern(leftmost);
+                            if (p == null)
+                            {
+                                return null;
+                            }
+
+                            var condition = (leftmost.Parent.Parent as BinaryExpressionSyntax)
+                                ?.WithLeft(((BinaryExpressionSyntax)leftmost.Parent).Right);
+
+                            return new Pattern.Guarded(p, (condition ?? node.Right).WalkDownParentheses());
+                        }
+
+                    default:
+                        return null;
+                }
+            }
+
+            StatementSyntax elseBody;
+            var sectionData = new List<(List<Pattern> patterns, StatementSyntax body)>();
+            var currentIfStatement = ifStatement;
+            while (true)
+            {
+                var operands = GetLogicalOrExpressionOperands(currentIfStatement.Condition).ToImmutableArray();
+                var patterns = new List<Pattern>(operands.Length);
+                foreach (var operand in operands)
+                {
+                    var pattern = GetPattern(operand);
+                    if (pattern == null)
+                    {
+                        return;
+                    }
+
+                    patterns.Insert(0, pattern);
+                }
+
+                sectionData.Add((patterns, currentIfStatement.Statement));
+                elseBody = currentIfStatement.Else?.Statement;
+                currentIfStatement = elseBody as IfStatementSyntax;
+                if (currentIfStatement == null)
+                {
+                    break;
+                }
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c =>
```

Comment:
would far prefer that the lambda be just a helper method.

basically, right now it's really hard to tell what's being captured and what state is flowing across reentrant calls.  By extracting out methods and being explicit about data flow, it's far easier ot reason about.